### PR TITLE
Custom exceptions inherit from AzureError

### PR DESCRIPTION
**Pull request by bradleydamato**
I used git grep "(Exception)" to search for custom exceptions that inherit from Exception in azure-sdk-for-python/sdk/eventhub that should be inheriting from AzureError.

I modified the class definitions for those exceptions to inherit from AzureError 

Following my changes, git grep yielded no results. 

I ran tox tests (see below)
image

### CODE REVIEW

File: sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py [hunk 1]
```diff
@@ -123,7 +125,7 @@ class OperationTimeoutError(EventHubError):
     """Operation timed out."""
 
 
-class OwnershipLostError(Exception):
+class OwnershipLostError(AzureError):
```
Comment (KieranBrantnerMagee) [inline sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:128]: @YijunXieMS Was there a historical reason this wasn't EventHubError?  I'm inclined to normalize but figured I should ask.

cc @yunhaoling (no action needed, just echoing the above as a scream-test)

File: sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py [hunk 2]
```diff
@@ -36,7 +38,7 @@ def _error_handler(error):
     return errors.ErrorAction(retry=True)
 
 
-class EventHubError(Exception):
+class EventHubError(AzureError):
```
Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:41]: As now the `EventHubError` inherits from the `AzureError`, I think we need to call the `super.__init__()` in the end of the constructor as following:

```
if details and isinstance(details, Exception):
    super(EventHubError, self).__init__(self.message, error=details)
else:
    super(EventHubError, self).__init__(self.message)
```

Comment (yunhaoling) [inline sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:41]: could you also kindly help update the changelog saying that:
```
**New features**
- `azure.eventhub.exceptions.EventHubError` now inherits from `azure.core.exceptions.AzureError`.
```

File: sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py (hunk 1 above)
Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:128]: @KieranBrantnerMagee , yea, technically, `OwnershipLostError` has nothing to do with the eventhub itself, it's related to load balance algorithm (claiming ownership) and blob operations. IIRC it's not used in the eventhub sdk, but used in the extensions sdks (checkpointstoreblob).

that's the reason I can remember why we didn't make it inherit from `EventHubError`.
but, I can also foresee the counter point on this "load balance, blob operation, they're all part of the event hub sdk, so it should be an EventHubError". With that being said, I now lean a bit towards making it inherit from `EventHubError`.

what's your thoughts on this?

File: sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py (hunk 2 above)
Comment (bradleydamato) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:41]: Hey @yunhaoling ; thanks for the review; do you mean that I should replace the current `__init__()` with something like this? 
```
def __init__(self, message, details=None):
        self.error = None
        self.message = message
        self.details = details
        if details and isinstance(details, Exception):
            try:
                condition = details.condition.value.decode("UTF-8")
            except AttributeError:
                try:
                    condition = details.condition.decode("UTF-8")
                except AttributeError:
                    condition = None
            if condition:
                _, _, self.error = condition.partition(":")
                self.message += "\nError: {}".format(self.error)
            try:
                self._parse_error(details.description)
                for detail in self.details:
                    self.message += "\n{}".format(detail)
            except:  # pylint: disable=bare-except
                self.message += "\n{}".format(details)
        if details and isinstance(details, Exception):
             super(EventHubError, self).__init__(self.message, error=details)
        else:
             super(EventHubError, self).__init__(self.message)
```


or should I replace it with this?
```
def __init__(self, message, details=None):
        self.error = None
        self.message = message
        self.details = details
        if details and isinstance(details, Exception):
              super(EventHubError, self).__init__(self.message, error=details)
        else:
              super(EventHubError, self).__init__(self.message)
        
```

Or should I replace it with something like this? 

```
def __init__(self, message, details=None):
        self.error = None
        self.message = message
        self.details = details
        if details and isinstance(details, AzureError):
              super(EventHubError, self).__init__(self.message, error=details)
        else:
              super(EventHubError, self).__init__(self.message)
```

Comment (bradleydamato) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:41]: Hey @yunhaoling , just wanted to ping this again. I'm not in a rush or anything, but also don't want this to get lost in a sea of notifs :)

Comment (bradleydamato) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:41]: @yunhaoling added in latest commit 👍

Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:41]: hey @bradleydamato , sorry for the late response.

I would expect it to be the first one, other approaches would break the message information completeness.

Comment (bradleydamato) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:41]: @yunhaoling no problem on the late response, pushing the commit now.

Comment (bradleydamato) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:41]: @yunhaoling unsure if the Azure Pipelines checks will auto-run again or not with my new commits. Is there a heuristic that would allow me to figure out whether they will run again or not? Either way, no rush on my end :).

File: sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py (hunk 1 above)
Comment (bradleydamato) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:131]: @yunhaoling @KieranBrantnerMagee just an FYI that the pylint error that ci is failing on below is related to this (I think):
[*pylint failure*
](https://dev.azure.com/azure-sdk/public/_build/results?buildId=578550&view=logs&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=23d29da4-00cb-5783-b17c-96bda25d9c52&l=77)
```
azure/eventhub/extensions/checkpointstoreblobaio/_blobstoragecsaio.py:189: [E1120(no-value-for-parameter), BlobCheckpointStore._claim_one_partition] No value for argument 'message' in constructor call
```
[line 189 in _blobstoragecasio.py](https://github.com/Azure/azure-sdk-for-python/blob/48ff8195424dbecbc738269ab85b9aa0072c56ab/sdk/eventhub/azure-eventhub-checkpointstoreblob-aio/azure/eventhub/extensions/checkpointstoreblobaio/_blobstoragecsaio.py#L189)
```raise OwnershipLostError()```

Not exactly sure what to do here, but if this is in fact related to the above conversation I'd be happy to modify my commits :).

Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py]: I'm sorry I completely miss the thread here.

it's because if `OwnershipLostError` inherits from `AzureError`, then `message` becomes a required parameter to the constructor.

I have updated your branch accordingly.

File: sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py (hunk 2 above)
Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:40]: you did great work, thanks!
there are two minor indentation issue, I've fixed them.

File: sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py (hunk 1 above)
Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py]: I think `OwnershipLostError` could inherit from `EventHubError`, the reason why we didn't do that before might be due to it's an error triggered by the storage-blob sdk, not by eventhub when failing to claim a partition (update a blob).

But conceptually this error is still related to the eventhub workflow and it would bring consistent error-handling experience if we could make it derived from `EventHubError`.

thoughts? @annatisch @rakshith91

File: sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py [hunk 3]
```diff
@@ -69,7 +70,10 @@ def __init__(self, message, details=None):
                     self.message += "\n{}".format(detail)
             except:  # pylint: disable=bare-except
                 self.message += "\n{}".format(details)
-        super(EventHubError, self).__init__(self.message)
+        if details and isinstance(details, Exception):
```
Comment (rakshith91) [inline sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:73]: what happens if  details is None? wondering if we really need this check

Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:73]: for clarification, are you saying that we should remove `details` check:
```python
if isinstance(details, Exception):
```
then I agree that `isinstance(details, Exception)` covers `None` check.


or the check `if details and isinstance(details, Exception):` should be completely removed?

Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py]: I have made the check a bit more compact, let me know how it looks to you

File: sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py (hunk 1 above)
Comment (annatisch) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py]: A couple of considerations here:
- What information is in this error? Is there any HTTP errors? For example - if we are wrapping a StorageErrorException from the Blobs SDK, this is raised on an HTTP error, and therefore we should also consider this an HTTP error. Conclusion: it should be an AzureError derivative.
- A customer might be broken if they have written code like the following. Is that a possible scenario?
```python
try:
    # run
except (EventHubError, AzureError):
    # do something
except OwnershipLostError:
    # do something else
```
- Our guidelines state that any custom errors should be based on those in `azure-core`:
https://azure.github.io/azure-sdk/python_implementation.html#python-errors-new-exceptions

Comment (annatisch) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py]: Additional note regarding the code snippet above:
- Is there any reason a customer would handle this error _differently_ to any other EventHubError? i.e. if customers have written a single `except EventHubError` clause - is there any reason they would not want this clause to be processing OwnershipLostErrors?

Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py]: @annatisch 
- currently the `OwnershipLostError` is wrapping `ResourceModifiedError` and  `ResourceExistsError` from the storage blob, see code [here](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/eventhub/azure-eventhub-checkpointstoreblob/azure/eventhub/extensions/checkpointstoreblob/_blobstoragecs.py#L163-L173), so yes it should inherit from `AzureError`.
- you're right, in this way it breaks customers code, but the good news is that we don't have `EventHubError` inheriting from `AzureError` before, so chances are small that users would catch `EventHubError` and `AzureError` simultaneously.
- This is a fairly good point which pushes me to think that users usually can't do anything except logging when they meet with the `OwnershipLostError`, meanwhile for other eventhub errors, users might be able to take some action.


To summarize, the thinking of how users deal with the error `OwnershipLostError` vs `EventHubError` makes me now lean towards `OwnershipLostError` inheriting from `AzureError`.

Comment (lmazuel) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:130]: Well, even that is breaking:
```python
try:
   # pass
except AzureError:
   # do something
except Exception:
   # do else
```

Since after adding `AzureError` to any exceptions, you change the path.

So I'm torn to the right approach.... @johanste ?

Comment (johanste) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:130]: Isn't this exception for the inner workings of the EPH? In other words, where would a user try to catch (`except`) this type?

Comment (yunhaoling) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:130]: User would catch this exception in the `on_error` callback if provided. This exception would be throwed during the [load-balancing stage](https://github.com/Azure/azure-sdk-for-python/blob/adfce6bcd0e66b9d02f2354c89b6fb94c7feb73f/sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/event_processor.py#L266-L286).

Comment (johanste) [reply sdk/eventhub/azure-eventhub/azure/eventhub/exceptions.py:130]: Catch or get it passed in?

**Comment by yunhaoling** [comment]
hey @bradleydamato , I need apologize to you that after discussion with the team we decided not to take the change. The major reasons are as follows:

- This would be a breaking change to customers who build their error handling on both EventHubError and AzureError, e.g. code like below would be broken by the change:
```python
try:
   # pass
except AzureError:
   # do something
except EventHubError:
   # do else
```
- The change not only impacts the `EventHubConsumerClient` where errors are being passed into callback, but also `EventHubProducerClient` where errors are catchable.
- Based on the reachability of the azure-eventhub SDK ([non-trivial downloading amount](https://pepy.tech/project/azure-eventhub)), this would potentially impact a large amount of users.

To conclude with, the disadvantages outweigh the advantages so we decide not to take the change.

We sincerely thank you for your contribution and it was really an excellent experience to collaborate with you on the project. You're always welcome by the azure-sdk team as an open source project contributor and we look forward to cooperation with you again in the future!

cc: @annatisch @lmazuel @johanste
